### PR TITLE
Added more filtering options to pgvector vectorstore

### PR DESCRIPTION
**Pull request by savoiepe**
- **Description:** Using PGVector vector store, it was only possible to filter for values equals, in or not in metadata. Extended this feature to work with the following keywords : IN, NIN, BETWEEN, GT, LT, NE, EQ, LIKE, CONTAINS, OR, AND

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 1]
```diff
@@ -505,22 +506,64 @@ def __query_collection(
 
             if filter is not None:
                 filter_clauses = []
-                IN, NIN = "in", "nin"
+                IN, NIN, BETWEEN, GT, LT, NE, EQ, LIKE, CONTAINS, OR, AND = "in", "nin", "between", "gt", "lt", "ne", "eq", "like", "contains", "or", "and"
+
+                def create_clause(value):
```
Comment (hwchase17) [inline libs/community/langchain_community/vectorstores/pgvector.py:511]: could this be its own method on the class? would make it more readable for sure

Comment (savoiepe) [reply libs/community/langchain_community/vectorstores/pgvector.py]: Yes of course :)

**Review by hwchase17 — APPROVED** [review]
thanks!

**Comment by stefanorosanelli** [comment]
Hi @savoiepe,

I was about to propose an almost identical PR, but yours is more complete. 
You saved me some work, thanks :wink: 

I suggest a slightly different version of `_create_filter_clause` method, see below. I would like to add `GE` and `LE` conditions and use `value_case_insensitive` for `if/elif` conditions. 

As a side note I think `filter: Optional[Dict[str, str]] = None` argument used in some `PGVector` methods should be  `filter: Optional[Dict[str, Union[str, Dict]]] = None` instead.


```python

    def _create_filter_clause(self, key, value):
        IN, NIN, BETWEEN, GE, GT, LE, LT, NE = "in", "nin", "between", "ge", "gt", "le", "lt", "ne"
        EQ, LIKE, CONTAINS, OR, AND = "eq", "like", "contains", "or", "and"

        value_case_insensitive = {
            k.lower(): v for k, v in value.items()
        }
        meta_key = self.EmbeddingStore.cmetadata[key]
        if IN in value_case_insensitive:
            filter_by_metadata = meta_key.astext.in_(value_case_insensitive[IN])
        elif NIN in value_case_insensitive:
            filter_by_metadata = meta_key.astext.not_in(value_case_insensitive[NIN])
        elif BETWEEN in value_case_insensitive:
            filter_by_metadata = meta_key.astext.between(
                str(value_case_insensitive[BETWEEN][0]),
                str(value_case_insensitive[BETWEEN][1])
                )
        elif GE in value_case_insensitive:
            filter_by_metadata = meta_key.astext >= str(value_case_insensitive[GE])
        elif GT in value_case_insensitive:
            filter_by_metadata = meta_key.astext > str(value_case_insensitive[GT])
        elif LE in value_case_insensitive:
            filter_by_metadata = meta_key.astext <= str(value_case_insensitive[LE])
        elif LT in value_case_insensitive:
            filter_by_metadata = meta_key.astext < str(value_case_insensitive[LT])
        elif NE in value_case_insensitive:
            filter_by_metadata = meta_key.astext != str(value_case_insensitive[NE])
        elif EQ in value_case_insensitive:
            filter_by_metadata = meta_key.astext == str(value_case_insensitive[EQ])
        elif LIKE in value_case_insensitive:
            filter_by_metadata = meta_key.astext.like(value_case_insensitive[LIKE])
        elif CONTAINS in value_case_insensitive:
            filter_by_metadata = meta_key.astext.contains(value_case_insensitive[CONTAINS])
        elif OR in value_case_insensitive:
            or_clauses = [
                self._create_filter_clause(key, sub_value) 
                for sub_value in value_case_insensitive[OR]
            ]
            filter_by_metadata = sqlalchemy.or_(or_clauses)
        elif AND in value_case_insensitive:
            and_clauses = [
                self._create_filter_clause(key, sub_value) 
                for sub_value in value_case_insensitive[AND]
            ]
            filter_by_metadata = sqlalchemy.and_(and_clauses)

        else:
            filter_by_metadata = None

        return filter_by_metadata
```

**Comment by rancomp** [comment]
Hey @savoiepe . It looks like this structure already exists in the repository. Specifically, `libs/langchain/chains/query_constructor/ir.py` has a `Comparator` class which contains all these names. Is it possible to use this class?

**Review by rancomp — COMMENTED** [review]
I feel like there's a way to simplify this code block. Can we use a dictionary/match statement to map the value to the method?
The object `self.EmbeddingStore.cmetadata[key]` is also repeated many times. Is it possible to hold it as a variable?

**Comment by baron** [comment]
would appreciate any attention to #11514 since this affects me using LC on a daily basis

**Comment by djo10** [comment]
@savoiepe can you please give an example of using AND, OR operators ?

**Comment by 765144989** [comment]
How should it be used in similarity_search()? Can you give some examples?

**Comment by 765144989** [comment]
> @savoiepe can you please give an example of using AND, OR operators ?

Have you found a way?

**Pull request merged**